# Board Game Insights - Data Exploration
Let the fun begin!

## Setup and Initialization

In [1]:
# import packages
import pandas as pd
import numpy as np


# Setup Constants
FILE_PATH = "bgg_dataset.csv"
BORDER = "\n" + "="*50 + "\n"

# Establish path
import pathlib
PATH = pathlib.Path(r"C:\Users\kyrst\OneDrive\Documents\DataAnalyticsPortfolio\bgg")

# Import CSV
df = pd.read_csv(PATH / FILE_PATH, sep=";") # sep is needed since this dataset separates columns with ;


# view sample
print(f"# Columns: {len(df.columns)}")
print(f"# Rows: {len(df)}")
print(f"# Unique Board Games: {df['ID'].nunique()}")
df.head()

# Columns: 14
# Rows: 20343
# Unique Board Games: 20327


,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
0,174430.0,Gloomhaven,2017.0,1,4,120,14,42055,"8,79",1,"3,86",68323.0,"Action Queue, Action Retrieval, Campaign / Bat...","Strategy Games, Thematic Games"
1,161936.0,Pandemic Legacy: Season 1,2015.0,2,4,60,13,41643,"8,61",2,"2,84",65294.0,"Action Points, Cooperative Game, Hand Manageme...","Strategy Games, Thematic Games"
2,224517.0,Brass: Birmingham,2018.0,2,4,120,14,19217,"8,66",3,"3,91",28785.0,"Hand Management, Income, Loans, Market, Networ...",Strategy Games
3,167791.0,Terraforming Mars,2016.0,1,5,120,12,64864,"8,43",4,"3,24",87099.0,"Card Drafting, Drafting, End Game Bonuses, Han...",Strategy Games
4,233078.0,Twilight Imperium: Fourth Edition,2017.0,3,6,480,14,13468,"8,70",5,"4,22",16831.0,"Action Drafting, Area Majority / Influence, Ar...","Strategy Games, Thematic Games"


## Data Cleaning

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20343 entries, 0 to 20342
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  20327 non-null  float64
 1   Name                20343 non-null  object 
 2   Year Published      20342 non-null  float64
 3   Min Players         20343 non-null  int64  
 4   Max Players         20343 non-null  int64  
 5   Play Time           20343 non-null  int64  
 6   Min Age             20343 non-null  int64  
 7   Users Rated         20343 non-null  int64  
 8   Rating Average      20343 non-null  object 
 9   BGG Rank            20343 non-null  int64  
 10  Complexity Average  20343 non-null  object 
 11  Owned Users         20320 non-null  float64
 12  Mechanics           18745 non-null  object 
 13  Domains             10184 non-null  object 
dtypes: float64(3), int64(6), object(5)
memory usage: 2.2+ MB


**Conclusions**
- `ID`, `Year Published` and `Owned Users` don't need to be floats
- `Rating Average` and `Complexity Average` use commas instead of decimal points, which makes pandas interpret them as objects
- `Name`, `Mechanics`, and `Domains` are objects that could be strings

We can fix some of this by importing the dataset again and establishing some rules up front!

In [3]:
df = pd.read_csv(
    PATH / FILE_PATH, 
    sep=";",            # Tells pandas to look for semicolons
    decimal=",",        # Fixes the '8,5' vs '8.5' issue automatically!
    engine="pyarrow",   # Uses the fast arrow engine for reading
    dtype={'ID':'string', 'Year Published': 'Int64', 'Owned Users': 'Int64'}, # Ensures that pyarrow doesn't make incorrect assumptions
    dtype_backend="numpy_nullable" # Safer middle ground for EDA
)

df.head()

,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
0,174430,Gloomhaven,2017,1,4,120,14,42055,8.79,1,3.86,68323,"Action Queue, Action Retrieval, Campaign / Bat...","Strategy Games, Thematic Games"
1,161936,Pandemic Legacy: Season 1,2015,2,4,60,13,41643,8.61,2,2.84,65294,"Action Points, Cooperative Game, Hand Manageme...","Strategy Games, Thematic Games"
2,224517,Brass: Birmingham,2018,2,4,120,14,19217,8.66,3,3.91,28785,"Hand Management, Income, Loans, Market, Networ...",Strategy Games
3,167791,Terraforming Mars,2016,1,5,120,12,64864,8.43,4,3.24,87099,"Card Drafting, Drafting, End Game Bonuses, Han...",Strategy Games
4,233078,Twilight Imperium: Fourth Edition,2017,3,6,480,14,13468,8.7,5,4.22,16831,"Action Drafting, Area Majority / Influence, Ar...","Strategy Games, Thematic Games"


**Confirm it worked:**

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20343 entries, 0 to 20342
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  20327 non-null  string 
 1   Name                20343 non-null  string 
 2   Year Published      20342 non-null  Int64  
 3   Min Players         20343 non-null  Int64  
 4   Max Players         20343 non-null  Int64  
 5   Play Time           20343 non-null  Int64  
 6   Min Age             20343 non-null  Int64  
 7   Users Rated         20343 non-null  Int64  
 8   Rating Average      20343 non-null  Float64
 9   BGG Rank            20343 non-null  Int64  
 10  Complexity Average  20343 non-null  Float64
 11  Owned Users         20320 non-null  Int64  
 12  Mechanics           18745 non-null  string 
 13  Domains             10184 non-null  string 
dtypes: Float64(2), Int64(8), string(4)
memory usage: 2.4 MB


### Null handling

In [5]:
# Count nulls per column
df_nulls = df.isnull().sum().sort_values(ascending=False)
df_nulls[df_nulls > 0]

Domains           10159
Mechanics          1598
Owned Users          23
ID                   16
Year Published        1
dtype: int64

In [6]:
# Inspect samples of the dataset where each column has nulls
from IPython.display import display

def inspect_nulls(df):
    # First, list the columns with nulls
    null_counts = df.isnull().sum()
    cols_with_nulls = null_counts[null_counts > 0].index

    # Add condition for no nulls
    if len(cols_with_nulls) == 0:
        print("Huzzah! No nulls found!")
        return

    # Loop for null cols
    for col in cols_with_nulls:
        print(BORDER)
        print(f" Top 5 rows where '{col}' is NULL")
        print(BORDER)
        display_df = df[df[col].isnull()].head()
        display(display_df)

inspect_nulls(df)



 Top 5 rows where 'ID' is NULL




,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
10776,<NA>,Ace of Aces: Jet Eagles,1990,2,2,20,10,110,6.26,10778,2.0,<NA>,<NA>,<NA>
10835,<NA>,Die Erben von Hoax,1999,3,8,45,12,137,6.05,10837,2.0,<NA>,<NA>,<NA>
11152,<NA>,Rommel in North Africa: The War in the Desert ...,1986,2,2,0,12,53,6.76,11154,4.0,<NA>,<NA>,<NA>
11669,<NA>,Migration: A Story of Generations,2012,2,4,30,12,49,7.2,11671,2.0,<NA>,<NA>,<NA>
12649,<NA>,Die Insel der steinernen Wachter,2009,2,4,120,12,49,6.73,12651,3.0,<NA>,<NA>,<NA>




 Top 5 rows where 'Year Published' is NULL




,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
13984,<NA>,Hus,<NA>,2,2,40,0,38,6.28,13986,2.0,<NA>,<NA>,<NA>




 Top 5 rows where 'Owned Users' is NULL




,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
2828,202755,Guildhall Fantasy: Fellowship,2016,2,4,45,10,565,7.13,2830,2.0,<NA>,"Hand Management, Take That, Set Collection",<NA>
3590,196305,Guildhall Fantasy: Alliance,2016,2,4,45,10,360,7.2,3592,2.14,<NA>,"Hand Management, Set Collection, Take That",<NA>
3739,196306,Guildhall Fantasy: Coalition,2016,2,4,45,10,336,7.19,3741,2.13,<NA>,"Hand Management, Set Collection, Take That",<NA>
5807,289,Chariot Lords,1999,3,4,360,12,221,6.68,5809,3.0,<NA>,"Area Movement, Variable Player Powers",<NA>
9202,6813,Operation Market Garden: Descent into Hell,1985,2,2,120,12,94,6.72,9204,3.0,<NA>,"Dice Rolling, Events, Grid Movement, Hexagon G...",<NA>




 Top 5 rows where 'Mechanics' is NULL




,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
1059,85256,Timeline: Inventions,2010,2,8,15,8,7257,6.71,1060,1.11,12448,<NA>,Family Games
1150,113401,Timeline: Events,2011,2,8,15,8,4208,6.78,1151,1.1,7924,<NA>,"Family Games, Party Games"
1216,131325,Timeline: Diversity,2012,2,8,15,8,3790,6.79,1217,1.07,7589,<NA>,"Family Games, Party Games"
1343,99975,Timeline: Discoveries,2011,2,8,15,8,3506,6.73,1344,1.18,6855,<NA>,Family Games
1397,145189,Timeline: Music & Cinema,2013,2,8,15,8,3090,6.75,1398,1.1,5986,<NA>,"Family Games, Party Games"




 Top 5 rows where 'Domains' is NULL




,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
663,293141,King of Tokyo: Dark Edition,2020,2,6,30,8,1854,7.89,664,1.63,5157,"Card Drafting, Dice Rolling, King of the Hill,...",<NA>
1011,177802,Smash Up: It's Your Fault!,2016,2,2,60,14,1663,7.53,1012,2.06,6347,"Area Majority / Influence, Hand Management, Ta...",<NA>
1047,226520,Exit: The Game - The Sinister Mansion,2018,1,4,90,12,1595,7.45,1048,2.61,3446,Cooperative Game,<NA>
1073,274638,Unmatched: Robin Hood vs. Bigfoot,2019,2,2,20,9,924,8.24,1074,1.93,2665,"Action Points, Hand Management, Line of Sight,...",<NA>
1092,215841,Exit: The Game - The Forgotten Island,2017,1,4,90,12,2736,7.02,1093,2.7,5438,Cooperative Game,<NA>


Based on the samples and counts, this is my conclusion for each column with missing values:
| Column Name | Null Count | Strategy |
| :--- | :--- | :--- |
| **Domains** | 10,159 | Fill with Null Indicator |
| **Mechanics** | 1,598 | Fill with Null Indicator |
| **Owned Users** | 23 | Drop Rows |
| **ID** | 16 | Drop Rows |
| **Year Published** | 1 | Drop Rows |

`Domains` and `Mechanics` have a lot of nulls. It could be random or the games with nulls could have more in common. Either way, let's track it by keeping these rows and imputing with something else.

Next, I'll explore `Domains` and `Mechanics` to understand them better before deciding on impute values and next steps

In [7]:
# Let's see how many unique values exist for Domains and Mechanis
check_unique = ['Domains', 'Mechanics']

for col in check_unique:
    print(f"{col}: {df[col].nunique()} unique values")


Domains: 39 unique values
Mechanics: 7381 unique values


Since `Domains` has a limited # of unique values, let's check them all out. 
- Conclusion: It's a mix of single value and comma delimited lists

In [8]:
df['Domains'].value_counts(dropna=False)

Domains
<NA>                                              10159
Wargames                                           3029
Strategy Games                                     1455
Family Games                                       1340
Abstract Games                                      869
Children's Games                                    708
Thematic Games                                      647
Party Games                                         409
Family Games, Strategy Games                        354
Customizable Games                                  235
Strategy Games, Thematic Games                      217
Family Games, Party Games                           139
Thematic Games, Wargames                            139
Abstract Games, Family Games                        116
Family Games, Thematic Games                        109
Children's Games, Family Games                      105
Strategy Games, Wargames                             99
Abstract Games, Strategy Games          

Since there are many unique values for `Mechanics`, let's look at the top and bottom values
- Conclusion: There are more lists, separated by a "," or a "/"

In [9]:
print(BORDER)
print("Top 10:")
print(BORDER)
print(df['Mechanics'].value_counts(dropna=False).head(10))
print(BORDER)
print("Bottom 10:")
print(BORDER)
print(df['Mechanics'].value_counts(dropna=False).tail(10))



Top 10:


Mechanics
<NA>                                      1598
Hand Management                            432
Hexagon Grid                               412
Dice Rolling                               372
Roll / Spin and Move                       369
Tile Placement                             285
Dice Rolling, Hexagon Grid, Simulation     260
Dice Rolling, Hexagon Grid                 256
Set Collection                             231
Hand Management, Set Collection            179
Name: count, dtype: Int64


Bottom 10:


Mechanics
Area Majority / Influence, Card Drafting, Modular Board                                                            1
Area Majority / Influence, Campaign / Battle Card Driven, Dice Rolling, Hand Management, Simulation, Tug of War    1
Area Movement, Dice Rolling, Events, Race, Scenario / Mission / Campaign Game, Simultaneous Action Selection       1
Area Majority / Influence, Dice Rolling, Push Your Luck, Take That                                        

Based on the investigation above, I'll simply fill nulls with 'Unknown' and then proceed with one-hot encoding later

In [10]:
# Drop rows based on nulls
drop_nulls = ['Owned Users', 'ID', 'Year Published']
df = df.dropna(subset = drop_nulls)

# Impute
df['Domains'] = df['Domains'].fillna("Unknown")
df['Mechanics'] = df['Mechanics'].fillna("Unknown")

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 20320 entries, 0 to 20342
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  20320 non-null  string 
 1   Name                20320 non-null  string 
 2   Year Published      20320 non-null  Int64  
 3   Min Players         20320 non-null  Int64  
 4   Max Players         20320 non-null  Int64  
 5   Play Time           20320 non-null  Int64  
 6   Min Age             20320 non-null  Int64  
 7   Users Rated         20320 non-null  Int64  
 8   Rating Average      20320 non-null  Float64
 9   BGG Rank            20320 non-null  Int64  
 10  Complexity Average  20320 non-null  Float64
 11  Owned Users         20320 non-null  Int64  
 12  Mechanics           20320 non-null  string 
 13  Domains             20320 non-null  string 
dtypes: Float64(2), Int64(8), string(4)
memory usage: 2.5 MB


In [11]:
df.head()

,ID,Name,Year Published,Min Players,Max Players,Play Time,Min Age,Users Rated,Rating Average,BGG Rank,Complexity Average,Owned Users,Mechanics,Domains
0,174430,Gloomhaven,2017,1,4,120,14,42055,8.79,1,3.86,68323,"Action Queue, Action Retrieval, Campaign / Bat...","Strategy Games, Thematic Games"
1,161936,Pandemic Legacy: Season 1,2015,2,4,60,13,41643,8.61,2,2.84,65294,"Action Points, Cooperative Game, Hand Manageme...","Strategy Games, Thematic Games"
2,224517,Brass: Birmingham,2018,2,4,120,14,19217,8.66,3,3.91,28785,"Hand Management, Income, Loans, Market, Networ...",Strategy Games
3,167791,Terraforming Mars,2016,1,5,120,12,64864,8.43,4,3.24,87099,"Card Drafting, Drafting, End Game Bonuses, Han...",Strategy Games
4,233078,Twilight Imperium: Fourth Edition,2017,3,6,480,14,13468,8.7,5,4.22,16831,"Action Drafting, Area Majority / Influence, Ar...","Strategy Games, Thematic Games"
